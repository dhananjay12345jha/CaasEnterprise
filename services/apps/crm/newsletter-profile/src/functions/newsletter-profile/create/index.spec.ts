import { faker } from "@faker-js/faker";
import {
  default as setNewsletterProfileMarketingPreferenceHandler,
  HandlerEvent,
} from "./index";
import { Context as LambdaContext } from "aws-lambda/handler";

const setMarketingPreferenceMock = jest.fn();

jest.mock("./create-newsletter-profile", () => {
  return {
    NewsletterProfile: jest.fn(() => {
      return {
        setMarketingPreference: setMarketingPreferenceMock,
      };
    }),
  };
});

describe("trigger set marketing preference of newsletter profile handler", () => {
  beforeEach(() => jest.restoreAllMocks());

  it("should throw an error if brand id is missing", async () => {
    const event: HandlerEvent = {
      version: "1.0",
      id: faker.datatype.string(),
      "detail-type": "newsletter.signup.request",
      time: faker.datatype.string(),
      detail: {
        referenced: false,
        metadata: { "x-emc-ubid": null },
        correlationId: faker.datatype.uuid(),
        payload: {
          email: faker.internet.email(),
          marketingPreference: "opt_in",
          timezone: faker.datatype.string(),
        },
      },
    };

    await expect(() =>
      setNewsletterProfileMarketingPreferenceHandler(event, {} as LambdaContext)
    ).rejects.toMatchSnapshot();
  });

  it("should return a confirmation response body and passes all input fields", async () => {
    setMarketingPreferenceMock.mockResolvedValue(true);

    const brandId = faker.datatype.uuid();

    const event: HandlerEvent = {
      version: "1.0",
      id: faker.datatype.uuid(),
      "detail-type": "newsletter.signup.request",
      time: faker.datatype.string(),
      detail: {
        referenced: false,
        metadata: { "x-emc-ubid": brandId },
        correlationId: faker.datatype.uuid(),
        payload: {
          email: faker.internet.email(),
          marketingPreference: "opt_in",
          timezone: "Europe/London",
        },
      },
    };

    await setNewsletterProfileMarketingPreferenceHandler(
      event,
      {} as LambdaContext
    );

    expect(setMarketingPreferenceMock).toBeCalledTimes(1);
    expect(setMarketingPreferenceMock).toBeCalledWith(
      brandId,
      event.detail.payload,
      event.time
    );
  });

  it("should throw an error when the setMarketingPreference throws an error", async () => {
    setMarketingPreferenceMock.mockRejectedValue(
      new Error("Error while triggering")
    );

    const brandId = faker.datatype.uuid();

    const event: HandlerEvent = {
      version: "1.0",
      id: faker.datatype.uuid(),
      "detail-type": "newsletter.signup.request",
      time: faker.datatype.string(),
      detail: {
        referenced: false,
        metadata: { "x-emc-ubid": brandId },
        correlationId: faker.datatype.uuid(),
        payload: {
          email: faker.internet.email(),
          marketingPreference: "opt_in",
          timezone: "Europe/London",
        },
      },
    };

    await expect(() =>
      setNewsletterProfileMarketingPreferenceHandler(event, {} as LambdaContext)
    ).rejects.toThrow(Error);
  });
});
