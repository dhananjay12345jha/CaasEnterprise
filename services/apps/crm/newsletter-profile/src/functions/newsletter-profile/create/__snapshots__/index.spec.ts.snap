// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`trigger set marketing preference of newsletter profile handler should throw an error if brand id is missing 1`] = `
Array [
  Object {
    "code": "invalid_type",
    "expected": "string",
    "message": "Expected string, received null",
    "path": Array [
      "detail",
      "metadata",
      "x-emc-ubid",
    ],
    "received": "null",
  },
  Object {
    "code": "custom",
    "message": "Invalid input",
    "path": Array [
      "detail",
      "payload",
      "timezone",
    ],
  },
]
`;
